### ISSUE TITLE
Cannot read properties of null (reading 'split') in exchangeCodeForSession

### ISSUE THREAD

**Issue by devbydaniel** [supabase/supabase#19896]
# Bug report

- [x] I confirm this is a bug with Supabase, not with my own application. 
- [x] I confirm I have searched the [Docs](https://docs.supabase.com), GitHub [Discussions](https://github.com/supabase/supabase/discussions), and [Discord](https://discord.supabase.com). 

## Describe the bug

I am seeing `TypeError - Cannot read properties of null (reading 'split')` in the `exchangeCodeForSession` method my `auth/callback` handler. I am running Next 14.0 with Supabase.

I am using the `ssr` package, version `0.0.4`. I've never had this error during development, so it's hard for me to tell the context in which the bug occurrs. 

## To Reproduce

Code for creating the client

```
export function createAuthHandlerClient(
  cookieStore: ReturnType<typeof cookies>
) {
  return createSupabaseServerClient(
    process.env.NEXT_PUBLIC_SUPABASE_URL!,
    process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
    {
      cookies: {
        get(name: string) {
          return cookieStore.get(name)?.value
        },
        set(name: string, value: string, options: CookieOptions) {
          try {
            cookieStore.set({ name, value, ...options })
          } catch (error) {
            // The `set` method was called from a Server Component.
            // This can be ignored if you have middleware refreshing
            // user sessions.
          }
        },
        remove(name: string, options: CookieOptions) {
          try {
            cookieStore.set({ name, value: '', ...options })
          } catch (error) {
            // The `delete` method was called from a Server Component.
            // This can be ignored if you have middleware refreshing
            // user sessions.
          }
        },
      },
    }
  )
}
```

code of `auth/callback` handler

```
import { NextResponse } from 'next/server'
import { cookies } from 'next/headers'
import { createAuthHandlerClient } from '@/adapter/_supabase/server'

export async function GET(request: Request) {
  // The `/auth/callback` route is required for the server-side auth flow implemented
  // by the Auth Helpers package. It exchanges an auth code for the user's session.
  // https://supabase.com/docs/guides/auth/auth-helpers/nextjs#managing-sign-in-with-code-exchange
  const requestUrl = new URL(request.url)
  const code = requestUrl.searchParams.get('code')

  if (code) {
    const cookieStore = cookies()
    const supabase = createAuthHandlerClient(cookieStore)
    await supabase.auth.exchangeCodeForSession(code)
  }

  const next = requestUrl.searchParams.get('next')
  if (next) {
    return NextResponse.redirect(
      `${process.env.NEXT_PUBLIC_BASEURL || requestUrl.origin}?redirect=${next}`
    )
  }
  return NextResponse.redirect(requestUrl.origin)
}
```

## Expected behavior

No errors on production.

## System information

```
"@supabase/ssr": "^0.0.4",
"@supabase/supabase-js": "^2.39.0", 
"next": "^14.0.3",
"react": "^18.2.0",
```

## Additional context

Any help would be greatly appreciated

**Comment by davegonzalez** [comment]
I'm also seeing this in a new Remix app with the following versions:
```
"@supabase/ssr": "^0.0.10",
"@supabase/supabase-js": "^2.39.0",
"@remix-run/express": "^2.2.0",
"@remix-run/node": "^2.2.0",
"@remix-run/react": "^2.2.0",
"react": "^18.2.0",
````
I have auth running server-side and have the following config options on `createServerClient`:
```
autoRefreshToken: false,
persistSession: false,
detectSessionInUrl: false
```

**Comment by notblong** [comment]
I had the same error previously and turned out for me is that the methods that I used to create supabaseClient for client and server are different (in client I used  method `createClient`  from '@supabase/supabase-js' and in server is from '@supabase/ssr').
I replaced my client method to `createBrowserClient` from the same lib '@supabase/ssr' and everything works fine, so I'm not sure is `createSupabaseServerClient` in your client side from '@supabase/ssr'

Ref: https://supabase.com/docs/guides/auth/server-side/creating-a-client?environment=client-component

**Comment by devbydaniel** [comment]
Thanks for your reply! I just realized I put the wrong code in my issue above. The client I am using in my auth handler to generate the client is the code in the [Route Handler tab](https://supabase.com/docs/guides/auth/server-side/creating-a-client?environment=route-handler) of the docs.

**Comment by caseybaggz** [comment]
Having the same issue here but I am consistently using the same SB APIs

**Comment by caseybaggz** [comment]
One of my customers said they got around the issue by doing a manual urldecodr and stripped off the proxy. Not sure what this means for SB but this is killing my release day. 😭

**Comment by rolandvar** [comment]
Good evening,
Currently some of my users are getting this error when they try to recover their password.
The strange thing is that it does not happen but only a few times.

Does anyone know what this error is caused by ?

We are using:
```
"next":"^14.0.4",
"@supabase/ssr":"^0.0.10",
"@supabase/supabase-js":"^2.39.2"
```

The frontend client is instantiated like this
```js
"use client";

import { createBrowserClient } from "@supabase/ssr";

export const broswerClient = () =>
  createBrowserClient(
    process.env.NEXT_PUBLIC_SUPABASE_URL!,
    process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!
  );
```

While the server one like this
```js
import { createServerClient, type CookieOptions } from "@supabase/ssr";
import { cookies } from "next/headers";

export const serverClient = (cookieStore: ReturnType<typeof cookies>) =>
  createServerClient(
    process.env.NEXT_PUBLIC_SUPABASE_URL!,
    process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
    {
      cookies: {
        get(name: string) {
          return cookieStore.get(name)?.value;
        },
        set(name: string, value: string, options: CookieOptions) {
          cookieStore.set({ name, value, ...options });
        },
        remove(name: string, options: CookieOptions) {
          cookieStore.set({ name, value: "", ...options });
        },
      },
    }
  );
```

Instead the route for the code exchange with the session is this
```js
import { NextResponse } from "next/server";
import { cookies } from "next/headers";

import { serverClient, Env } from "@/utils";

export const runtime = "edge";
export const dynamic = "force-dynamic";

export async function GET(request: Request) {
  // The `/auth/callback` route is required for the server-side auth flow implemented
  // by the Auth Helpers package. It exchanges an auth code for the user's session.
  // https://supabase.com/docs/guides/auth/auth-helpers/nextjs#managing-sign-in-with-code-exchange
  const requestUrl = new URL(request.url);
  const code = requestUrl.searchParams.get("code");

  if (code) {
    const cookieStore = cookies();
    const supabase = serverClient(cookieStore);
    await supabase.auth.exchangeCodeForSession(code);
  }

  // URL to redirect to after sign in process completes
  const next = requestUrl.searchParams.get("next");
  const nextUrl = next ? new URL(next, Env.APP_HOSTNAME) : requestUrl.origin;

  return NextResponse.redirect(nextUrl);
}
```

Finally, if anyone would like to try, a URL that causes the error is this
```/api/auth/recovery?code=67fb24e3-158f-43fe-95f7-5bbbfefbeb20&next=%2Fauth%2Fupdate-password```

Any suggestions are appreciated, thank you all.

Best regards

**Comment by rolandvar** [comment]
> ```js
> exchangeCodeForSession
> ```

Hello, everyone,
After a long search I found that my problem is related to the fact that I was sending the password reset request via one device and opening the email link on another device.
As documented [here ](https://github.com/supabase/auth-helpers/issues/545#issuecomment-1568655226)it is not possible to start the password recovery flow on one device and complete it on another because exchangeCodeForSession needs a cookie that is set when you make the password reset request.

Consequently, as advised by [this comment](https://github.com/supabase/auth-helpers/issues/545#issuecomment-1703071442), I modified my flow to use PKCE with verifyOtp and adapted the email template.

Now everything seems to work as it should.

I hope to be of help to someone.

Best regards
